File: lib/ex_gix/object_id.ex
defmodule ExGix.ObjectId do
  @moduledoc """
  A representation of an object hash, most commonly Sha1.
  """

  @type t :: reference()
  @type kind :: :sha1 | :sha256

  @doc """
  Create an instance from a buffer of 40 bytes or 64 bytes encoded with hexadecimal notation.

  ## Examples

      {:ok, oid} = ExGix.ObjectId.from_hex("e69de29bb2d1d6434b8b29ae775ad8c2e48c5391")
      is_reference(oid) #=> true

  """
  @spec from_hex(String.t()) :: {:ok, t()} | {:error, String.t()}
  def from_hex(hex) when is_binary(hex) do
    ExGix.Native.object_id_from_hex(hex)
  end

  @doc """
  Return a type which displays this oid as hex in full.

  ## Examples

      {:ok, oid} = ExGix.ObjectId.empty_tree()
      ExGix.ObjectId.to_hex(oid)
      #=> "4b825dc642cb6eb9a060e54bf8d69288fbee4904"

  """
  @spec to_hex(t()) :: String.t()
  def to_hex(object_id) when is_reference(object_id) do
    ExGix.Native.object_id_to_hex(object_id)
  end

  @doc """
  The kind of hash used for this instance.
  """
  @spec kind(t()) :: kind()
  def kind(object_id) when is_reference(object_id) do
    ExGix.Native.object_id_kind(object_id)
  end

  @doc """
  The hash of an empty blob.

  ## Examples

      {:ok, oid} = ExGix.ObjectId.empty_blob()
      ExGix.ObjectId.is_empty_blob(oid) #=> true

  """
  @spec empty_blob(kind()) :: {:ok, t()} | {:error, String.t()}
  def empty_blob(kind \\ :sha1) when is_atom(kind) do
    ExGix.Native.object_id_empty_blob(kind)
  end

  @doc """
  The hash of an empty tree.

  ## Examples

      {:ok, oid} = ExGix.ObjectId.empty_tree()
      ExGix.ObjectId.is_empty_tree(oid) #=> true

  """
  @spec empty_tree(kind()) :: {:ok, t()} | {:error, String.t()}
  def empty_tree(kind \\ :sha1) when is_atom(kind) do
    ExGix.Native.object_id_empty_tree(kind)
  end

  @doc """
  Returns an instances whose bytes are all zero.

  ## Examples

      {:ok, oid} = ExGix.ObjectId.null()
      ExGix.ObjectId.is_null(oid) #=> true

  """
  @spec null(kind()) :: {:ok, t()} | {:error, String.t()}
  def null(kind \\ :sha1) when is_atom(kind) do
    ExGix.Native.object_id_null(kind)
  end

  @doc """
  Returns true if this hash consists of all null bytes.
  """
  @spec is_null(t()) :: boolean()
  def is_null(object_id) when is_reference(object_id) do
    ExGix.Native.object_id_is_null(object_id)
  end

  @doc """
  Returns true if this hash is equal to an empty blob.
  """
  @spec is_empty_blob(t()) :: boolean()
  def is_empty_blob(object_id) when is_reference(object_id) do
    ExGix.Native.object_id_is_empty_blob(object_id)
  end

  @doc """
  Returns true if this hash is equal to an empty tree.
  """
  @spec is_empty_tree(t()) :: boolean()
  def is_empty_tree(object_id) when is_reference(object_id) do
    ExGix.Native.object_id_is_empty_tree(object_id)
  end
end

File: native/ex_gix/src/object_id.rs
use rustler::ResourceArc;

pub struct ObjectIdResource(pub gix::ObjectId);

rustler::atoms! {
    sha1,
    sha256,
}

fn atom_to_kind(atom: rustler::Atom) -> Result<gix::hash::Kind, String> {
    if atom == sha1() {
        Ok(gix::hash::Kind::Sha1)
    } else {
        Err(format!("Unsupported hash kind"))
    }
}

fn kind_to_atom(kind: gix::hash::Kind) -> rustler::Atom {
    match kind {
        gix::hash::Kind::Sha1 => sha1(),
        _ => sha1(), // default to sha1
    }
}

#[rustler::nif]
pub fn object_id_from_hex(hex: String) -> Result<ResourceArc<ObjectIdResource>, String> {
    match gix::ObjectId::from_hex(hex.as_bytes()) {
        Ok(id) => Ok(ResourceArc::new(ObjectIdResource(id))),
        Err(e) => Err(e.to_string()),
    }
}

#[rustler::nif]
pub fn object_id_to_hex(id: ResourceArc<ObjectIdResource>) -> String {
    id.0.to_hex().to_string()
}

#[rustler::nif]
pub fn object_id_kind(id: ResourceArc<ObjectIdResource>) -> rustler::Atom {
    kind_to_atom(id.0.kind())
}

#[rustler::nif]
pub fn object_id_empty_blob(kind: rustler::Atom) -> Result<ResourceArc<ObjectIdResource>, String> {
    let k = atom_to_kind(kind)?;
    Ok(ResourceArc::new(ObjectIdResource(
        gix::ObjectId::empty_blob(k),
    )))
}

#[rustler::nif]
pub fn object_id_empty_tree(kind: rustler::Atom) -> Result<ResourceArc<ObjectIdResource>, String> {
    let k = atom_to_kind(kind)?;
    Ok(ResourceArc::new(ObjectIdResource(
        gix::ObjectId::empty_tree(k),
    )))
}

#[rustler::nif]
pub fn object_id_null(kind: rustler::Atom) -> Result<ResourceArc<ObjectIdResource>, String> {
    let k = atom_to_kind(kind)?;
    Ok(ResourceArc::new(ObjectIdResource(gix::ObjectId::null(k))))
}

#[rustler::nif]
pub fn object_id_is_null(id: ResourceArc<ObjectIdResource>) -> bool {
    id.0.is_null()
}

#[rustler::nif]
pub fn object_id_is_empty_blob(id: ResourceArc<ObjectIdResource>) -> bool {
    id.0.is_empty_blob()
}

#[rustler::nif]
pub fn object_id_is_empty_tree(id: ResourceArc<ObjectIdResource>) -> bool {
    id.0.is_empty_tree()
}
